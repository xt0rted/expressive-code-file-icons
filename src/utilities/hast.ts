import { addClassName, getClassNames } from "@expressive-code/core/hast";

import type { Element } from "@expressive-code/core/hast";

export function isTerminal(element: Element) {
  const classes = getClassNames(element);

  return classes.includes("is-terminal");
}

export function addClassNames(element: Element, classNames?: string[]) {
  if (!classNames) {
    return;
  }

  for (const className of classNames) {
    addClassName(element, className);
  }
}
