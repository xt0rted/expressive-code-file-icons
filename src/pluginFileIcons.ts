import { select, setProperty } from "@expressive-code/core/hast";

import { addClassNames, isTerminal } from "./utilities/hast.js";
import { iconForFile } from "./utilities/icons.js";

import type { ExpressiveCodePlugin } from "@expressive-code/core";

import type { IconName } from "./iconNames.js";
import type { PluginFileIconsOptions } from "./options.js";

export function pluginFileIcons({ iconClass, titleClass }: PluginFileIconsOptions = {}): ExpressiveCodePlugin {
  return {
    name: "File icons",
    hooks: {
      preprocessMetadata({ codeBlock }) {
        const { metaOptions, props } = codeBlock;

        props.icon = (metaOptions.getString("icon") as IconName) ?? props.icon;
        props.noIcon = metaOptions.getBoolean("no-icon") ?? props.noIcon;
      },
      postprocessRenderedBlock: async ({ codeBlock, renderData }) => {
        if (codeBlock.props.noIcon) {
          return;
        }

        if (isTerminal(renderData.blockAst)) {
          return;
        }

        const titleText = codeBlock.props["title"]; // Comes from the frames plugin

        if (!titleText) {
          return;
        }

        const iconSvgElement = await iconForFile(titleText, codeBlock.language, codeBlock.props.icon);

        if (!iconSvgElement) {
          return;
        }

        setProperty(iconSvgElement, "aria-hidden", "true");

        addClassNames(iconSvgElement, iconClass?.split(" "));

        const title = select("figcaption > .title", renderData.blockAst);

        if (!title) {
          return;
        }

        title.children?.unshift(iconSvgElement);

        addClassNames(title, titleClass?.split(" "));
      },
    },
  };
}
