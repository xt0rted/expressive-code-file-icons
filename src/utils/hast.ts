import { getClassNames } from "@expressive-code/core/hast";

import type { Element } from "@expressive-code/core/hast";

export function isTerminal(element: Element) {
  const classes = getClassNames(element);

  return classes.includes("is-terminal");
}
