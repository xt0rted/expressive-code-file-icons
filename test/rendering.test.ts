import { pluginFrames } from "@expressive-code/plugin-frames";
import { describe, test } from "vitest";

import { pluginFileIcons } from "../src/pluginFileIcons.js";
import { debugSettings, pluginDebugStyles } from "./helpers/debugging.js";
import { selectSingle, titleSelector } from "./helpers/hast.js";
import { buildFixture, renderAndOutputHtmlSnapshot } from "./helpers/testing.js";
import { testName } from "./helpers/vitest.js";

import type { ExpressiveCodePlugin } from "@expressive-code/core";

describe("renders an icon in the title", async () => {
  const plugins: ExpressiveCodePlugin[] = [pluginDebugStyles(), pluginFrames(), pluginFileIcons(debugSettings)];

  test.concurrent("adds icon based on code block language", async ({ expect, task }) => {
    await renderAndOutputHtmlSnapshot({
      testName: testName(task),
      testBaseDir: import.meta.dirname,
      fixture: buildFixture({
        code: "",
        language: "js",
        meta: 'title="test config"',
        plugins,
        blockValidationFn: (block) => {
          const title = selectSingle(titleSelector, block.renderedGroupAst);

          expect(title).toMatchSnapshot();
        },
      }),
    });
  });

  test.concurrent("adds icon based on title file extension", async ({ expect, task }) => {
    await renderAndOutputHtmlSnapshot({
      testName: testName(task),
      testBaseDir: import.meta.dirname,
      fixture: buildFixture({
        code: "",
        language: "js",
        meta: 'title="test.ts"',
        plugins,
        blockValidationFn: (block) => {
          const title = selectSingle(titleSelector, block.renderedGroupAst);

          expect(title).toMatchSnapshot();
        },
      }),
    });
  });

  test.concurrent("adds icon based on title file name", async ({ expect, task }) => {
    await renderAndOutputHtmlSnapshot({
      testName: testName(task),
      testBaseDir: import.meta.dirname,
      fixture: buildFixture({
        code: "",
        language: "js",
        meta: 'title="astro.config.js"',
        plugins,
        blockValidationFn: (block) => {
          const title = selectSingle(titleSelector, block.renderedGroupAst);

          expect(title).toMatchSnapshot();
        },
      }),
    });
  });
});
