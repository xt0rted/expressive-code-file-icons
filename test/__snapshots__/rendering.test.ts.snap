// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders an icon in the title > adds icon based on code block language 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 80,
                  "line": 1,
                  "offset": 79,
                },
                "start": {
                  "column": 68,
                  "line": 1,
                  "offset": 67,
                },
              },
              "type": "text",
              "value": "file_type_js",
            },
          ],
          "position": {
            "end": {
              "column": 88,
              "line": 1,
              "offset": 87,
            },
            "start": {
              "column": 61,
              "line": 1,
              "offset": 60,
            },
          },
          "properties": {},
          "tagName": "title",
          "type": "element",
        },
        {
          "children": [],
          "position": {
            "end": {
              "column": 680,
              "line": 1,
              "offset": 679,
            },
            "start": {
              "column": 88,
              "line": 1,
              "offset": 87,
            },
          },
          "properties": {
            "d": "M18.774,19.7a3.727,3.727,0,0,0,3.376,2.078c1.418,0,2.324-.709,2.324-1.688,0-1.173-.931-1.589-2.491-2.272l-.856-.367c-2.469-1.052-4.11-2.37-4.11-5.156,0-2.567,1.956-4.52,5.012-4.52A5.058,5.058,0,0,1,26.9,10.52l-2.665,1.711a2.327,2.327,0,0,0-2.2-1.467,1.489,1.489,0,0,0-1.638,1.467c0,1.027.636,1.442,2.1,2.078l.856.366c2.908,1.247,4.549,2.518,4.549,5.376,0,3.081-2.42,4.769-5.671,4.769a6.575,6.575,0,0,1-6.236-3.5ZM6.686,20c.538.954,1.027,1.76,2.2,1.76,1.124,0,1.834-.44,1.834-2.15V7.975h3.422V19.658c0,3.543-2.078,5.156-5.11,5.156A5.312,5.312,0,0,1,3.9,21.688Z",
            "style": "fill:#f5de19",
          },
          "tagName": "path",
          "type": "element",
        },
      ],
      "position": {
        "end": {
          "column": 686,
          "line": 1,
          "offset": 685,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": {
        "aria-hidden": "true",
        "className": [
          "size-4",
        ],
        "viewBox": "0 0 32 32",
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "tagName": "svg",
      "type": "element",
    },
    {
      "type": "text",
      "value": "test config",
    },
  ],
  "properties": {
    "className": [
      "title",
      "flex",
      "items-center",
      "gap-1",
    ],
  },
  "tagName": "span",
  "type": "element",
}
`;

exports[`renders an icon in the title > adds icon based on title file extension 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 88,
                  "line": 1,
                  "offset": 87,
                },
                "start": {
                  "column": 68,
                  "line": 1,
                  "offset": 67,
                },
              },
              "type": "text",
              "value": "file_type_typescript",
            },
          ],
          "position": {
            "end": {
              "column": 96,
              "line": 1,
              "offset": 95,
            },
            "start": {
              "column": 61,
              "line": 1,
              "offset": 60,
            },
          },
          "properties": {},
          "tagName": "title",
          "type": "element",
        },
        {
          "children": [],
          "position": {
            "end": {
              "column": 1010,
              "line": 1,
              "offset": 1009,
            },
            "start": {
              "column": 96,
              "line": 1,
              "offset": 95,
            },
          },
          "properties": {
            "d": "M23.827,8.243A4.424,4.424,0,0,1,26.05,9.524a5.853,5.853,0,0,1,.852,1.143c.011.045-1.534,1.083-2.471,1.662-.034.023-.169-.124-.322-.35a2.014,2.014,0,0,0-1.67-1c-1.077-.074-1.771.49-1.766,1.433a1.3,1.3,0,0,0,.153.666c.237.49.677.784,2.059,1.383,2.544,1.095,3.636,1.817,4.31,2.843a5.158,5.158,0,0,1,.416,4.333,4.764,4.764,0,0,1-3.932,2.815,10.9,10.9,0,0,1-2.708-.028,6.531,6.531,0,0,1-3.616-1.884,6.278,6.278,0,0,1-.926-1.371,2.655,2.655,0,0,1,.327-.208c.158-.09.756-.434,1.32-.761L19.1,19.6l.214.312a4.771,4.771,0,0,0,1.35,1.292,3.3,3.3,0,0,0,3.458-.175,1.545,1.545,0,0,0,.2-1.974c-.276-.395-.84-.727-2.443-1.422a8.8,8.8,0,0,1-3.349-2.055,4.687,4.687,0,0,1-.976-1.777,7.116,7.116,0,0,1-.062-2.268,4.332,4.332,0,0,1,3.644-3.374A9,9,0,0,1,23.827,8.243ZM15.484,9.726l.011,1.454h-4.63V24.328H7.6V11.183H2.97V9.755A13.986,13.986,0,0,1,3.01,8.289c.017-.023,2.832-.034,6.245-.028l6.211.017Z",
            "style": "fill:#007acc",
          },
          "tagName": "path",
          "type": "element",
        },
      ],
      "position": {
        "end": {
          "column": 1016,
          "line": 1,
          "offset": 1015,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": {
        "aria-hidden": "true",
        "className": [
          "size-4",
        ],
        "viewBox": "0 0 32 32",
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "tagName": "svg",
      "type": "element",
    },
    {
      "type": "text",
      "value": "test.ts",
    },
  ],
  "properties": {
    "className": [
      "title",
      "flex",
      "items-center",
      "gap-1",
    ],
  },
  "tagName": "span",
  "type": "element",
}
`;

exports[`renders an icon in the title > adds icon based on title file name 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 89,
                  "line": 1,
                  "offset": 88,
                },
                "start": {
                  "column": 68,
                  "line": 1,
                  "offset": 67,
                },
              },
              "type": "text",
              "value": "file_type_astroconfig",
            },
          ],
          "position": {
            "end": {
              "column": 97,
              "line": 1,
              "offset": 96,
            },
            "start": {
              "column": 61,
              "line": 1,
              "offset": 60,
            },
          },
          "properties": {},
          "tagName": "title",
          "type": "element",
        },
        {
          "children": [],
          "position": {
            "end": {
              "column": 529,
              "line": 1,
              "offset": 528,
            },
            "start": {
              "column": 97,
              "line": 1,
              "offset": 96,
            },
          },
          "properties": {
            "d": "M11.025 20.499c-.532 1.75-.154 4.184 1.105 5.331v-.042l.042-.112c.154-.741.756-1.203 1.526-1.175.713.014 1.12.392 1.217 1.217.042.308.042.616.056.938v.098c0 .7.196 1.371.588 1.959.35.56.84.993 1.497 1.287l-.028-.056-.028-.112c-.49-1.469-.14-2.49 1.147-3.358l.392-.266.868-.573a4.254 4.254 0 0 0 1.791-3.037c.07-.532 0-1.05-.154-1.553l-.21.14c-1.945 1.035-4.17 1.4-6.325.98-1.301-.197-2.56-.56-3.498-1.652z",
            "fill": "url(#a)",
          },
          "tagName": "path",
          "type": "element",
        },
        {
          "children": [],
          "position": {
            "end": {
              "column": 750,
              "line": 1,
              "offset": 749,
            },
            "start": {
              "column": 529,
              "line": 1,
              "offset": 528,
            },
          },
          "properties": {
            "d": "M4.925 20.191s3.736-1.82 7.486-1.82l2.84-8.759c.098-.42.406-.7.756-.7s.644.28.756.714l2.826 8.746c4.45 0 7.487 1.82 7.487 1.82L20.709 2.84c-.168-.518-.49-.84-.896-.84h-7.612c-.406 0-.7.322-.896.84z",
            "fill": "#fff",
          },
          "tagName": "path",
          "type": "element",
        },
        {
          "children": [],
          "position": {
            "end": {
              "column": 1639,
              "line": 1,
              "offset": 1638,
            },
            "start": {
              "column": 750,
              "line": 1,
              "offset": 749,
            },
          },
          "properties": {
            "d": "M25.684,27.193l.456-.447c2.112.068,2.144,0,2.237-.219l.58-1.393L29,25l-.047-.115c-.025-.061-.1-.243-1.42-1.483v-.652c1.521-1.445,1.489-1.519,1.4-1.73l-.577-1.407c-.087-.211-.12-.294-2.216-.237l-.457-.465a10,10,0,0,0-.071-2.053l-.059-.132-1.508-.65c-.222-.1-.3-.136-1.724,1.393l-.64-.009c-1.466-1.543-1.535-1.515-1.755-1.428l-1.4.559c-.219.088-.3.119-.212,2.219l-.453.445c-2.11-.068-2.142.006-2.234.219l-.581,1.393L15,21l.048.116c.025.06.1.24,1.419,1.481v.65c-1.521,1.445-1.488,1.519-1.4,1.731l.577,1.409c.089.215.12.292,2.216.238l.456.467a9.967,9.967,0,0,0,.07,2.05l.058.133,1.518.654c.221.091.3.124,1.717-1.4l.64.008c1.468,1.545,1.543,1.515,1.757,1.43l1.4-.558C25.7,29.324,25.773,29.294,25.684,27.193Zm-6.013-3.262a2.467,2.467,0,0,1,.623-2.691,2.451,2.451,0,0,1,.826-.509,2.526,2.526,0,0,1,3.246,1.429,2.466,2.466,0,0,1-1.405,3.184,2.553,2.553,0,0,1-3.29-1.413Z",
            "fill": "#99b8c4",
          },
          "tagName": "path",
          "type": "element",
        },
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "position": {
                    "end": {
                      "column": 1836,
                      "line": 1,
                      "offset": 1835,
                    },
                    "start": {
                      "column": 1797,
                      "line": 1,
                      "offset": 1796,
                    },
                  },
                  "properties": {
                    "offset": "0",
                    "stopColor": "#D83333",
                  },
                  "tagName": "stop",
                  "type": "element",
                },
                {
                  "children": [],
                  "position": {
                    "end": {
                      "column": 1875,
                      "line": 1,
                      "offset": 1874,
                    },
                    "start": {
                      "column": 1836,
                      "line": 1,
                      "offset": 1835,
                    },
                  },
                  "properties": {
                    "offset": "1",
                    "stopColor": "#F041FF",
                  },
                  "tagName": "stop",
                  "type": "element",
                },
              ],
              "position": {
                "end": {
                  "column": 1892,
                  "line": 1,
                  "offset": 1891,
                },
                "start": {
                  "column": 1645,
                  "line": 1,
                  "offset": 1644,
                },
              },
              "properties": {
                "gradientTransform": "translate(-.673 -2.198) scale(1.3993)",
                "gradientUnits": "userSpaceOnUse",
                "id": "a",
                "x1": "8.19",
                "x2": "16.91",
                "y1": "23",
                "y2": "18.89",
              },
              "tagName": "linearGradient",
              "type": "element",
            },
          ],
          "position": {
            "end": {
              "column": 1899,
              "line": 1,
              "offset": 1898,
            },
            "start": {
              "column": 1639,
              "line": 1,
              "offset": 1638,
            },
          },
          "properties": {},
          "tagName": "defs",
          "type": "element",
        },
      ],
      "position": {
        "end": {
          "column": 1905,
          "line": 1,
          "offset": 1904,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "properties": {
        "aria-hidden": "true",
        "className": [
          "size-4",
        ],
        "viewBox": "0 0 32 32",
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "tagName": "svg",
      "type": "element",
    },
    {
      "type": "text",
      "value": "astro.config.js",
    },
  ],
  "properties": {
    "className": [
      "title",
      "flex",
      "items-center",
      "gap-1",
    ],
  },
  "tagName": "span",
  "type": "element",
}
`;
